torch_audiomentations/augmentations/noise_gates.py

```diff
@@ -0,0 +1,128 @@
+import torch
+import numpy as np
+from torch import Tensor
+from typing import Optional
+from torchaudio.functional import DB_to_amplitude,amplitude_to_DB
+
+from ..core.transforms_interface import BaseWaveformTransform
+from torch_audiomentations.utils.object_dict import ObjectDict
+
+
+
+class SpectralGating(BaseWaveformTransform):
+
+    supported_modes = {"per_batch", "per_example", "per_channel"}
+
+    def __init__(
+        self,
+        std_away = 1.0,
+        n_grad_freq=2,
+        n_grad_time=4,
+        q = 0.5,
+        decrease_prop=1,
+        n_fft=2048,
+        win_length=2048,
+        hop_length=512,
+        mode: str = "per_example",
+        p: float = 0.5,
+        p_mode: Optional[str] = None,
+        sample_rate: Optional[int] = None,
+        target_rate: Optional[int] = None,
+        output_type: Optional[str] = None,
+    ):
+        super(SpectralGating, self).__init__(
+            mode=mode,
+            p=p,
+            p_mode=p_mode,
+            sample_rate=sample_rate,
+            target_rate=target_rate,
+            output_type=output_type,
+        )
+        self.n_fft = n_fft
+        self.win_length = win_length
+        self.hop_length = hop_length
+        self.decrease_prop = min(1.0,decrease_prop)  ##max decrease prop
+        self.q = q
+        self.n_grad_freq = n_grad_freq
+        self.n_grad_time = n_grad_time
+
+
+
+    def randomize_parameters(
+        self,
+        samples: Tensor = None,
+        sample_rate: Optional[int] = None,
+        targets: Optional[Tensor] = None,
+        target_rate: Optional[int] = None):
+
+        dist = torch.distributions.Uniform(0.5 ,self.decrease_prop)
+        self.transform_parameters['decrease_prop'] = dist.sample((samples.shape[0],))
+
+        
+
+    def apply_transform(
+        self,
+        samples: Tensor = None,
+        sample_rate: Optional[int] = None,
+        targets: Optional[Tensor] = None,
+        target_rate: Optional[int] = None) -> ObjectDict:
+
+        ### threshold 
+        if self.mode == "per_batch":
```

Comment:
What was the motivation for having separate code for per_batch, per_example and per_channel? Normally the BaseWaveformTransform will take care of reshaping the audio tensor to a "per example" kind so the transform doesn't have to deal with each mode separately.

Reply:
To calculate the noise threshold differently for each mode. For example, if mode = "per_example", the noise threshold is set differently for each example and processed accordingly. You can see the code for each one is a bit different from others.

Reply:
I see. But I get the feeling that the mode is taken into acount "twice", as in it gets handled first by the base class (reshaping) and then also in this transform. I haven't studied the code in detail though.

The base class will reshape the audio tensor. For example:
* per_example: `(num_examples, num_channels, samples)`
* per_batch: `(1, num_examples * num_channels, samples)`
* per_channel: `num_examples * num_channels, 1, samples)`

So the idea is that the transform can always assume a "per_example" way of transforming the audio for simplicity

Reply:
@iver56 got it. I have made the changes to remove mode-based processing. I have added the function to the demo as well. Let me know what you think.